ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs

```diff
@@ -45,9 +48,128 @@ import           Ouroboros.Consensus.HardFork.Combinator.Ledger.Query
+  -> NP f xs
+  -> Encoding
+encodeNodeToClientNP
+  encodeElement
+  (HardForkCodecConfig (PerEraCodecConfig ccfgs))
+  version
+  xs
+  = case version of
+      HardForkNodeToClientDisabled versionX -> case ccfgs of
+        ccfg :* _ -> case xs of
+          x :* _ -> encodeElement ccfg versionX x
+      HardForkNodeToClientEnabled _ subVersions ->
+        let components :: [Encoding]
+            components = catMaybes
+              $ hcollapse
+              $ hczipWith3
+                  (Proxy @SerialiseConstraintsHFC)
+                  (\ccfg subVersionMay x -> K $ case subVersionMay of
+                    EraNodeToClientEnabled subVersion -> Just (encodeElement ccfg subVersion x)
+                    -- Omit disabled eras
+                    EraNodeToClientDisabled           -> Nothing
+                  )
+                  ccfgs
+                  subVersions
+                  xs
+            listLen = fromIntegral (length components)
+        in Enc.encodeListLen listLen <> mconcat components
+
+-- | Decoding of `NP f xs`. If any eras are disabled in the version and hence
+-- missing in the serialisation, then this throws an exception. In effect,
+-- deserialisation of product types is only supported when the sender uses an
+-- equal or superset of eras.
+decodeNodeToClientNP ::
+     forall xs f. SerialiseHFC xs
+  => (   forall x. SerialiseConstraintsHFC x
+      => CodecConfig x
+      -> BlockNodeToClientVersion x
+      -> (forall s. Decoder s (f x))
+     )
+  -- ^ The decoding of the individual elements (assuming era `x` is enabled).
+  -> CodecConfig (HardForkBlock xs)
+  -> HardForkNodeToClientVersion xs
+  -> (forall s. Decoder s (NP f xs))
+decodeNodeToClientNP
+  decodeElement
+  (HardForkCodecConfig (PerEraCodecConfig ccfgs))
+  version
+  = case version of
+      HardForkNodeToClientDisabled versionX -> case ccfgs of
+        (ccfg :* Nil) -> do
+          singleElement <- decodeElement ccfg versionX
+          return (singleElement :* Nil)
+        _ -> failVersion
+
+      HardForkNodeToClientEnabled _ subVersions -> do
+        enforceSize failVersionTxt expectedN
+        hsequence' $ hczipWith
+          (Proxy @SerialiseConstraintsHFC)
+          (\ccfg subVersionMay -> Comp $ case subVersionMay of
+              EraNodeToClientEnabled subVersion -> decodeElement ccfg subVersion
+              -- Fail if any era is disabled
+              EraNodeToClientDisabled           -> failVersion
+          )
+          ccfgs
+          subVersions
+    where
+      expectedN = lengthSList (Proxy @xs)
+
+      failVersion :: Decoder s a
+      failVersion = fail failVersionStr
+      failVersionStr = "decodeNodeToClient: (NP f xs): incompatible node-to-client version"
+      failVersionTxt = T.pack failVersionStr
 
 instance SerialiseHFC xs => SerialiseNodeToClientConstraints (HardForkBlock xs)
 
+instance SerialiseHFC xs => SerialiseNodeToClient (HardForkBlock xs) (HardForkLedgerConfig xs) where
```

Comment:
[This same comment applies to all of the new instances in this module.]

Style: the `=` ends the first line, it doesn't start the second line. (See other instances that were already in this file, eg.)